### PR TITLE
RFC, support scion for HSDP/TP/CP, etc 

### PR DESCRIPTION
**Pull request by rakkit**

### CODE REVIEW

File: torchtitan/components/ft.py
Comment (rakkit) [inline torchtitan/components/ft.py:1]: This branch (FTParallelDims) is removed from the mainstream. After rebase w,e will not bother with `ft.FTParallelDims`

File: torchtitan/components/metrics.py
Comment (rakkit) [inline torchtitan/components/metrics.py:1]: just formating

File: torchtitan/components/optimizer.py
Comment (rakkit) [inline torchtitan/components/optimizer.py:1]: I moved the norm logger to a separate helper file, so the optimizer.py file remains clean. Later, I may also clean the extract_param_group here, but for now, all looks good.

This is a special (distribution) norm log moved to dist-scion

File: torchtitan/components/optimizer.py [hunk 1]
```diff
@@ -314,164 +213,30 @@ def load_state_dict(self, state_dict: dict[str, Any]) -> None:
...
-                g = p.grad
-
-            assert isinstance(g, DTensor), "Expected gradient to be a DTensor"
-            return g.redistribute(placements=[Replicate()] * g.device_mesh.ndim)
-        else:
-            raise TypeError(
-                f"Optimizer {optimizer.__class__.__name__} does not support "
-                f"gradient computation."
-            )
+    def calculate_norm_at_next_step(self):
```
Comment (rakkit) [inline torchtitan/components/optimizer.py:217]: for dist-scion.     tell Scion to calculate norm when trigger's optim.step' and put norm into the dict

File: torchtitan/components/optimizer.py [hunk 2]
```diff
@@ -314,164 +213,30 @@ def load_state_dict(self, state_dict: dict[str, Any]) -> None:
...
-                        "backend_steps": group["backend_steps"],
-                    }
-                elif isinstance(optimizer, (torch.optim.Adam, torch.optim.AdamW)):
-                    param_kwargs = {
-                        "eps": group["eps"],
-                        "betas": group["betas"],
-                        "weight_decay": group["weight_decay"],
-                    }
+                if isinstance(optimizer, DistributedScion):
+                    all_norms.update(optimizer.get_norms_at_current_step())
```
Comment (rakkit) [inline torchtitan/components/optimizer.py:232]: for dist-scion.     we directly took norm's results.

else-wise, for other optimizer, we need to manually calculate them (with the legacy implementation)

File: torchtitan/components/optimizer.py [hunk 3]
```diff
@@ -630,10 +396,19 @@ def build_optimizers(
         job_config (JobConfig): Job config containing the optimizer name and parameters.
     """
     optim_in_bwd = job_config.optimizer.early_step_in_backward
-    if optim_in_bwd and job_config.parallelism.pipeline_parallel_degree > 1:
-        raise NotImplementedError(
-            "Optimizers in backward is not supported with pipeline parallelism."
-        )
+    if optim_in_bwd:
```
Comment (rakkit) [inline torchtitan/components/optimizer.py:400]: these tooked from upstream, dont know what does it means

File: torchtitan/components/optimizer.py [hunk 4]
```diff
@@ -741,7 +513,10 @@ def build_optimizers(
 
         param_groups_config.append(param_group_config)
 
-    optimizer_kwargs["extra_kwargs"] = extra_kwargs
+    optimizer_kwargs["extra_kwargs"] = {
```
Comment (rakkit) [inline torchtitan/components/optimizer.py:538]: upstream now directly pass `parallel_dims`,

File: torchtitan/config_manager.py [hunk 5]
```diff
@@ -295,6 +295,34 @@ def __init__(self):
                 Exponent applied to the final output layer's input dimensionality
                 to obtain its init std factor.""",
         )
+        self.parser.add_argument(
```
Comment (rakkit) [inline torchtitan/config_manager.py:312]: for depth-scaling ablation

File: torchtitan/config_manager.py [hunk 6]
```diff
@@ -384,7 +415,7 @@ def __init__(self):
             "--optimizer.zeropower_backend",
             type=str,
             default="newtonschulz5",
-            choices=["newtonschulz5", "svd"],
+            choices=["newtonschulz5", "svd", "polar_express"],
```
Comment (rakkit) [inline torchtitan/config_manager.py]: new backend (different coeff)

File: torchtitan/distributed/parallel_dims.py
Comment (rakkit) [inline torchtitan/distributed/parallel_dims.py:1]: tooked from upstream

File: torchtitan/distributed/pipeline.py
Comment (rakkit) [inline torchtitan/distributed/pipeline.py:1]: from upstream.  [disable MTP x PP for now]

File: torchtitan/models/MoEllama/loss.py
Comment (rakkit) [inline torchtitan/models/MoEllama/loss.py:1]: this file seems not used at all?

File: torchtitan/models/inits.py
Comment (rakkit) [inline torchtitan/models/inits.py:1]: remove the werid wrapper. make it cleaner.

fix the orthogonal that supports the initialization of the hybrid placement tensor.  (have 2 ways not, will ablate later see which one is more stable)

File: torchtitan/models/llama3/infra/parallelize.py
Comment (rakkit) [inline torchtitan/models/llama3/infra/parallelize.py:1]: take some changes from up stream.

and wrap up TP for dealing with 'norm-everywhere`

File: torchtitan/optimizers/norm_helper.py
Comment (rakkit) [inline torchtitan/optimizers/norm_helper.py:1]: add fused norms,
mainly merge the RMS-RMS norm and condition number computation together.  can save about 50% compute time for each matrix.

**Comment by janEbert** [comment]
Most of this had already been merged. Also merged the pipeline hacks. Let me know if anything is missing, e.g., like the MidNorm stuff that I got lost.
